File: frontend/src/app/services/reservation/eventReservation.service.ts
import { Injectable } from "@angular/core";
import { HttpClient } from "@angular/common/http";
import { Routes } from "../../config/routes.config";
import { Observable } from "rxjs/Observable";
import { EventReservation } from "../../models/EventReservation";

/**
 * Az eseményre vonatkozó foglalásokhoz tartozó service osztály
 */
@Injectable()
export class EventReservationService {
  constructor(private http: HttpClient) {}

  getAccepted(): Observable<EventReservation[]> {
    return <Observable<EventReservation[]>>(
      this.http.get(Routes.getUrl(Routes.EVENT_RESERVATION_GET_ACCEPTED))
    );
  }

  findByUsername(username: string): Observable<EventReservation[]> {
    return <Observable<EventReservation[]>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_FIND_BY_USERNAME) +
          "/" +
          username
      )
    );
  }

  findByStatus(status: string): Observable<EventReservation[]> {
    return <Observable<EventReservation[]>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_FIND_BY_STATUS + "/" + status)
      )
    );
  }

  findByBuildingAndClassroom(
    building: string,
    classroom: string
  ): Observable<EventReservation[]> {
    return <Observable<EventReservation[]>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_FIND_BY_BUILDING_AND_CLASSROOM) +
          `?building=${building}&classroom=${classroom}`
      )
    );
  }

  findByName(name: string): Observable<EventReservation[]> {
    return <Observable<EventReservation[]>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_FIND_BY_NAME) + "/" + name
      )
    );
  }

  setStatus(id: number, status: string): Observable<EventReservation> {
    return <Observable<EventReservation>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_SET_STATUS) +
          `?id=${id}&status=${status}`
      )
    );
  }

  createEventReservation(
    reservation: EventReservation
  ): Observable<EventReservation> {
    return <Observable<EventReservation>>(
      this.http.post(
        Routes.getUrl(Routes.EVENT_RESERVATION_CREATE_RESERVATION),
        reservation
      )
    );
  }

  deleteByUsername(username: string): Observable<any> {
    return <Observable<any>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_DELETE_BY_USERNAME) +
          "/" +
          username
      )
    );
  }

  deleteByBuildingAndClassroom(
    building: string,
    classroom: string
  ): Observable<any> {
    return <Observable<any>>(
      this.http.get(
        Routes.getUrl(
          Routes.EVENT_RESERVATION_DELETE_BY_BUILDING_AND_CLASSROOM
        ) + `?building=${building}&classroom=${classroom}`
      )
    );
  }

  deleteByStatus(status: string): Observable<any> {
    return <Observable<any>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_DELETE_BY_STATUS) +
          `?status=${status}`
      )
    );
  }

  deleteByName(name: string): Observable<any> {
    return <Observable<any>>(
      this.http.get(
        Routes.getUrl(Routes.EVENT_RESERVATION_DELETE_BY_NAME) + `?name=${name}`
      )
    );
  }
}

File: rrs-server/src/main/java/RoomReservationSystem/service/EventReservationService.java
package RoomReservationSystem.service;

import RoomReservationSystem.dto.EventReservationDTO;
import RoomReservationSystem.exception.BuildingNotExistsException;
import RoomReservationSystem.exception.ClassroomNotExistsException;
import RoomReservationSystem.exception.EventReservationAlredyExistsException;
import RoomReservationSystem.exception.EventReservationNotExistsException;
import RoomReservationSystem.exception.SemesterNotExistsException;
import RoomReservationSystem.exception.StatusNotExistsException;
import RoomReservationSystem.exception.UserNotExistsException;
import RoomReservationSystem.model.EventReservation;
import java.util.List;

/**
 * Az eseményre vonatkozó foglalásokkal kapcsolatos műveletekért felelős
 * interfész Részletes információ a függvényekről a megválósításnál
 *
 * @author Tomecz Patrik
 */
public interface EventReservationService {

    EventReservation save(EventReservationDTO reservation) throws UserNotExistsException, ClassroomNotExistsException, StatusNotExistsException, SemesterNotExistsException, BuildingNotExistsException, EventReservationAlredyExistsException;

    EventReservation findById(int id) throws EventReservationNotExistsException;

    EventReservation setStatus(int id, String status) throws StatusNotExistsException, EventReservationNotExistsException;

    EventReservation findByName(String name) throws EventReservationNotExistsException;

    EventReservation update(int id, EventReservationDTO eventReservationDTO) throws EventReservationNotExistsException, UserNotExistsException, ClassroomNotExistsException, StatusNotExistsException, SemesterNotExistsException, BuildingNotExistsException;

    List<EventReservation> getAll();

    List<EventReservation> findByUsername(String username) throws UserNotExistsException;

    List<EventReservation> findByStatus(String statusName) throws StatusNotExistsException;

    List<EventReservation> findByBuildingAndClassroom(String building, String classroom) throws ClassroomNotExistsException, BuildingNotExistsException;

    List<String> getNames();

    boolean existsById(int id);

    boolean existsByName(String name);
    
    void deleteById(int id) throws EventReservationNotExistsException;

    void deleteByUsername(String username) throws UserNotExistsException;

    void deleteByBuildingAndClassroom(String building, String classroom) throws ClassroomNotExistsException, BuildingNotExistsException;

    void deleteByStatus(String status) throws StatusNotExistsException;

    void deleteByName(String name) throws EventReservationNotExistsException;
}
